### ISSUE TITLE
Trying to reproduce monodepth2

### ISSUE THREAD

**Issue by zshn25** [jspenmar/monodepth_benchmark#4]
I am trying to reproduce monodepth2 results using the benchmark. I run

```bash
python api/train/train.py -cfg-file cfg/abl_bb/resnet18.yaml --cfg-default cfg/abl_bb/default.yaml --ckpt-dir logs/benchmark/ --name md2 --version 42 --seed 42
```

where `resnet18.yaml` is

```yaml
net:
  depth:
    enc_name: 'resnet18'

dataset:
  type: 'kitti'
  supp_idxs: [ -1, 1]

loader:
  batch_size: 36

```

But, when I evaluate it, my best model is not as good at monodepth2

```yaml
Ratio: 26.717182107430105
AbsRel: 0.1205853946784904
SqRel: 0.9113600682814631
RMSE: 4.980470443558659
LogRMSE: 0.19857460770280325
$\delta < 1.25$: 0.8655597941752403
$\delta < 1.25^2$: 0.9562950148720958
$\delta < 1.25^3$: 0.9799390811001284

```

**Comment by zshn25** [comment]
@jspenmar could you please verify if I’m using the correct config as intended?

Thank you

**Comment by jspenmar** [comment]
Sorry for the late reply, I must have missed the first email. 

I'm assuming that these results are on the base Kitti Eigen? The results already look quite close to those in the original Monodepth2 paper, but I would recommend experimenting with some of the hyperparameters if you want to tune it further. For instance, when increasing the batch size by a factor of `N`, it is typically recommended to also increase the learning rate by the same factor.  Finally, it is worth training a few models using the same config and a range of different random seeds, as the model performances can vary quite a bit across runs.

Other than that, the config looks okay, assuming that you want to train a purely monocular model. Please note that the image reconstruction loss config should also be 
```yaml
loss:
  img_recon:
    use_min: True
    use_automask: True
```

Hope this helps!

**Comment by zshn25** [comment]
Thank you for the suggestions. I've tried increasing the lr with the same factor as the batch size and ran a couple of other seeds but without success. One run were worse than the previous result, the other lead to instability and the network hadn't learn anything.


The best I could get was by running with the same batch size and lr as you, 8 and 1e-4 resp. which gave

```yaml
Ratio: 26.208320740818806
AbsRel: 0.11708547223033317
SqRel: 0.8824378384044728
RMSE: 4.825611422359516
LogRMSE: 0.19479319161416125
$\delta < 1.25$: 0.8734269378929359
$\delta < 1.25^2$: 0.9585055564659608
$\delta < 1.25^3$: 0.9807919814269638
Chamfer: 0.8874067592312992
F-Score: 22.079211754128444
IoU: 12.631402063917053
F-Score-20: 49.25878780507289
IoU-20: 33.23689125942192
```
which is closer to what monodepth2 reports but still not quite there.

Is there a way to save Top-n models and not just the best one? I suspect that by tracking AbsRel of the val split and saving it's best leads to only saving the overfitted model and maybe the best one is somewhere in Top-5.

Also, why is just changing the seed leading to very different results?

**Comment by jspenmar** [comment]
For reference, could you post exact which results from the Monodepth2 paper you are trying to reproduce. As far as I can tell, you're referring to the middle result in this row, as it's the fully monocular model. If so, I'd say that the difference at this point is minimal and could be caused by many different factors, such as different pytorch versions, pretrained models and random seeds. 

![Screenshot 2023-01-24 at 10 00 41](https://user-images.githubusercontent.com/43554262/214262949-f063f7c9-a49f-486d-a34d-7d9a6b81c37f.png)

In terms of saving multiple checkpoints, pytorch lightning seems to [provide this capability](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.ModelCheckpoint.html#pytorch_lightning.callbacks.ModelCheckpoint.params.save_top_k). Note that we also always save the `last` checkpoint, so you could try evaluating that one. 

Finally, the difference from the seeds is expected. Different seeds results in different initialization and ordering of the data, which can result in vastly different results. A lot of pytorch/CUDA operations are also not deterministic. That is why we propose training multiple models and reporting the average performance over them.
